R/move_parsing_utils.R

```diff
@@ -0,0 +1,227 @@
+#####################################################################
+# This file contains functions for parsing moves output by Maculay2 #
+#####################################################################
+
+
+#' Filter raw moves
+#' 
+#' Filter out moves from the raw move data.frame. 
+#' 
+#' Applies the following filters to the raw moves file:
+#' \itemize{
+#'    \item Remove rows whose elements do not sum to zero
+#'    \item Remove moves that add or subtract a non-existant edge
+#' }
+#' 
+#' @param raw data.frame
+#' 
+#' @return data.frame
+filter_raw_moves <- function(raw) {
+  raw[rowSums(raw) == 0 & rowSums(abs(raw[, get_indices("none")])) == 0, ]
+}
+
+
+#' Row index offset
+index_offset <- function(type) {
+  switch(
+    type,
+    "none" = 1,
+    "out" = 2,
+    "in" = 3,
+    "bi" = 4
+  )
+}
+
+# TODO: make work with other sized graphs
+#' Indices of edge states
+#' 
+#' Get the indices of the raw move vector corresponding to an edge state. 
+#' 
+#' Each edge can be in one of 4 states. The moves output by Maculay2 represent 
+#' these in chunks of 4  within a vector, one chunk for each edge. Each state
+#' has an offset within the chunk (see \code{\link{index_offset}}. The particular state
+#' of the \emph{i}th edge can be accessed with the index \emph{offset + 4*(i-1)}.
+#' 
+#' @param type character 
+#' @param n numeric, the number of edges in the graph
+#' 
+#' @return numeric vector of indices
+get_indices <- function(type, n=12) {
+  offset <- index_offset(type)
+  i <- seq(1:n)
+  return(offset + 4 * (i - 1))
+}
+
+
+#' Split a move into red and blue parts
+split_move <- function(move_row) {
+  
+  r_row <- move_row
+  b_row <- move_row
+  r_row[which(r_row == 1)] <- 0
+  b_row[which(b_row == -1)] <- 0
+  
+  return(list(r = abs(r_row), b = abs(b_row)))
+}
+
+
+# TODO: make work with other sized graphs
+# convert move to adjacency matrix
+convert_move_to_adj <- function(move, symmetric=FALSE) {
+
+  stopifnot(length(move) == 4 * 3)
+
+  m <- matrix(move, 4, 3, byrow = TRUE)
+  a <- matrix(0, 4, 4)
+  for (i in 1:4) {
+    a[i, ] <- append(m[i, ], 0, i - 1)
+  }
+  if (isFALSE(symmetric)) {
+    return(a)
+  } else {
+    return(a + t(a))
+  }
+}
+
+#' Get adjaceny matrices of a directed move
+#' 
+#' @param move_row numeric vector
+#' 
+#' @return list of matrices, one for each part of the move
+get_directed_move_matrix <- function(move_row) {
+  
+  out_moves <- split_move(move_row[get_indices("out")])
+  out_moves_m <- lapply(X = out_moves, FUN = convert_move_to_adj)
+
+  in_moves <- split_move(move_row[get_indices("in")])
+  in_moves_m <- lapply(X = in_moves, FUN = convert_move_to_adj)
+
```

Comment:
The "in"s should be transposed, but this passed because of a mistake in the unit test.